crypto/pkcs7/pk7_lib.c

```diff
@@ -367,7 +367,9 @@ int PKCS7_SIGNER_INFO_set(PKCS7_SIGNER_INFO *p7i, X509 *x509, EVP_PKEY *pkey,
                          V_ASN1_NULL, NULL))
         return 0;
 
-    if (EVP_PKEY_is_a(pkey, "EC") || EVP_PKEY_is_a(pkey, "DSA"))
+    if (EVP_PKEY_is_a(pkey, "EC")
+            || EVP_PKEY_is_a(pkey, "DSA")
+            || EVP_PKEY_is_a(pkey, "SM2"))
```

Comment:
I don't like this approach. Hard coding algorithms is a bad idea.

Reply:
It seems that it would be a good way to use switch to check the result of calling `EVP_PKEY_get_id(pkey)`.

Reply:
I would actually rather see a new function `pkcs7_sign_verify_setup` that's algorithm agnostic.  But that requires some other changes to fully support fetching signature algorithms by algorithm identifier (including direct OIDs).